single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,875 @@
+            )
+            raise UpgradeInProgressError
+
+        try:
+            # Adds the newly added/updated units.
+            self.mongo_manager.process_added_units()
+        except (NotReadyError, PyMongoError) as e:
+            logger.error(f"Not reconfiguring: error={e}")
+            self.charm.status_manager.to_waiting("waiting to reconfigure replica set")
+            raise
+
+    @override
+    def on_secret_changed(self, secret_label: str, secret_id: str) -> None:
+        """Handles secrets changes event.
+
+        When user run set-password action, juju leader changes the password inside the database
+        and inside the secret object. This action runs the restart for monitoring tool and
+        for backup tool on non-leader units to keep them working with MongoDB. The same workflow
+        occurs on TLS certs change.
+        """
+        if (
+            generate_secret_label(self.charm.app.name, self.charm.peer_rel_name, Scope.APP)
+            == secret_label
+        ):
+            scope = Scope.APP
+        elif (
+            generate_secret_label(self.charm.app.name, self.charm.peer_rel_name, Scope.UNIT)
+            == secret_label
+        ):
+            scope = Scope.UNIT
+        else:
+            logging.debug("Secret %s changed, but it's unknown", secret_id)
+            return
+        logging.debug("Secret %s for scope %s changed, refreshing", secret_id, scope)
+        self.state.secrets.get(scope)
+
+        # Always update the PBM and mongodb exporter configuration so that if
+        # the secret changed, the configuration is updated and will still work
+        # afterwards.
+        self.mongodb_exporter_config_manager.configure_and_restart()
+        self.backup_manager.configure_and_restart()
+
+        # Always process the statuses.
+        self.charm.status_manager.process_and_share_statuses()
+
+    @override
+    def on_relation_departed(self, departing_unit: Unit | None) -> None:
+        """Handles the relation departed events."""
+        if not self.charm.unit.is_leader() or departing_unit == self.charm.unit:
+            return
+        if self.state.upgrade_in_progress:
+            # do not defer or return here, if a user removes a unit, the config will be incorrect
+            # and lead to MongoDB reporting that the replica set is unhealthy, we should make an
+            # attempt to fix the replica set configuration even if an upgrade is occurring.
+            logger.warning(
+                "Removing replicas during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
+            )
+        self.update_hosts()
+        self.charm.status_manager.process_and_share_statuses()
+
+    @override
+    def on_storage_attached(self) -> None:  # pragma: nocover
+        """Handler for `storage_attached` event.
+
+        This should handle fixing the permissions for the data dir.
+        """
+        if self.substrate == Substrates.VM:
```

Comment:
nit: return if K8s (but again if thats not your style its fine) its just useful for reducing indentation